From 8f33c9173c52b3c577afdb7ac97bddeb6f6c90ef Mon Sep 17 00:00:00 2001
From: Joel Jakobsson <joel@compiler.org>
Date: Thu, 16 Feb 2023 22:17:44 +0100
Subject: [PATCH 1/3] Add test to detect changes to the Numeric format written
 to disk

This test is only intended to demonstrate the patch, and is not meant
to be kept in HEAD, since the disk format depends on platform endianess.
---
 src/test/regress/expected/numeric.out | 41 +++++++++++++++++++++++++++
 src/test/regress/sql/numeric.sql      | 32 +++++++++++++++++++++
 2 files changed, 73 insertions(+)

diff --git a/src/test/regress/expected/numeric.out b/src/test/regress/expected/numeric.out
index 56a3f3630a..e77bf267d0 100644
--- a/src/test/regress/expected/numeric.out
+++ b/src/test/regress/expected/numeric.out
@@ -3528,3 +3528,44 @@ SELECT pg_lsn(18446744073709551616::numeric);
 ERROR:  pg_lsn out of range
 SELECT pg_lsn('NaN'::numeric);
 ERROR:  cannot convert NaN to pg_lsn
+--
+-- Detect numeric disk format changes
+--
+CREATE TABLE num_test_disk_format
+(
+  c1 numeric, c2 numeric, c3 numeric, c4 numeric,
+  c5 numeric, c6 numeric, c7 numeric, c8 numeric,
+  c9 numeric, c10 numeric, c11 numeric, c12 numeric
+);
+INSERT INTO num_test_disk_format
+(
+   c1, c2, c3, c4,
+   c5, c6, c7, c8,
+   c9, c10, c11, c12
+)
+VALUES
+(
+    0, 1, 2, 3,
+    255, 256, 9999, (10000 * 0xab + 0xcd)::numeric * 10000 + 0xef,
+    4, 5, 6, 7
+);
+CHECKPOINT;
+SELECT regexp_replace(
+    right(encode(pg_read_binary_file(pg_relation_filepath('num_test_disk_format')),'hex'),16*8),
+    '([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})',
+    '\1 \2 \3 \4 \5 \6 \7 \8'||E'\n',
+    'g'
+);
+     regexp_replace
+-------------------------
+ 07 00 80 0b 00 80 01 00+
+ 0b 00 80 02 00 0b 00 80+
+ 03 00 0b 00 80 ff 00 0b+
+ 00 80 00 01 0b 00 80 0f+
+ 27 13 02 80 ab 00 cd 00+
+ ef 00 0b 00 80 04 00 0b+
+ 00 80 05 00 0b 00 80 06+
+ 00 0b 00 80 07 00 00 00+
+
+(1 row)
+
diff --git a/src/test/regress/sql/numeric.sql b/src/test/regress/sql/numeric.sql
index 2db7656e84..8a6b8d941b 100644
--- a/src/test/regress/sql/numeric.sql
+++ b/src/test/regress/sql/numeric.sql
@@ -1476,3 +1476,35 @@ SELECT pg_lsn(18446744073709551615::numeric);
 SELECT pg_lsn(-1::numeric);
 SELECT pg_lsn(18446744073709551616::numeric);
 SELECT pg_lsn('NaN'::numeric);
+
+--
+-- Detect numeric disk format changes
+--
+CREATE TABLE num_test_disk_format
+(
+  c1 numeric, c2 numeric, c3 numeric, c4 numeric,
+  c5 numeric, c6 numeric, c7 numeric, c8 numeric,
+  c9 numeric, c10 numeric, c11 numeric, c12 numeric
+);
+
+INSERT INTO num_test_disk_format
+(
+   c1, c2, c3, c4,
+   c5, c6, c7, c8,
+   c9, c10, c11, c12
+)
+VALUES
+(
+    0, 1, 2, 3,
+    255, 256, 9999, (10000 * 0xab + 0xcd)::numeric * 10000 + 0xef,
+    4, 5, 6, 7
+);
+
+CHECKPOINT;
+
+SELECT regexp_replace(
+    right(encode(pg_read_binary_file(pg_relation_filepath('num_test_disk_format')),'hex'),16*8),
+    '([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})',
+    '\1 \2 \3 \4 \5 \6 \7 \8'||E'\n',
+    'g'
+);
--
2.37.1 (Apple Git-137.1)